huntsman/observatory.py

```diff
@@ -472,43 +355,217 @@ def _create_autoguider(self):
+                else:
+                    n_good_exposures[cam_name] += 1
+
+                # Check if we have enough good flats for this camera
+                if n_good_exposures[cam_name] >= max_num_exposures:
+                    self.logger.debug(f'Enough flats taken for {cam_name}.')
+                    finished[cam_name] = True
+                    continue
+
+                # Calculate next exposure time
+                elapsed_time = (utils.current_time() - start_time).sec
+                next_exptime = self._autoflat_next_exptime(
+                        current_exptime, elapsed_time, target_counts, mean_counts)
+                self.logger.debug(f'Suggested exposure time for {cam_name}: {next_exptime}.')
+
+                # Check the next exposure time is within limits
+                if next_exptime > max_exptime:
+                    self.logger.debug(f'Suggested exposure time for {cam_name}'
+                                      f' is too long: {next_exptime}.')
+                    if not self.past_midnight:
+                        finished[cam_name] = True  # Its getting darker, so finish
+                        continue
+                    next_exptime = max_exptime
+                elif next_exptime < min_exptime:
+                    self.logger.debug(f'Suggested exposure time for {cam_name}'
+                                      f' is too short: {next_exptime}.')
+                    if self.past_midnight:
+                        finished[cam_name] = True  # Its getting lighter, so finish
+                        continue
+                    next_exptime = min_exptime
+
+                # Update the next exposure time
+                exptimes[cam_name].append(next_exptime)
+
+            # Check if all the exposures in this loop are too bright
+            if self.past_midnight:
+                if all_too_faint:
+                    self.logger.debug('All exposures are too faint. Waiting 30 seconds...')
+                    time.sleep(30)
+            else:
+                if all_too_bright:
+                    self.logger.debug('All exposures are too bright. Waiting 30 seconds...')
+                    time.sleep(30)
+
+        # Return the exposure times
+        return exptimes
+
+    def _autoflat_mean_counts(self, filename, bias, min_counts=1):
+        """ Read the data and calculate a clipped-mean count rate.
 
-        self.logger.debug("Flat-field observation: {}".format(flat_obs))
+        Args:
+            filename (str): The filename containing the data.
+            bias (float): The bias level to subtract from the image.
+            min_counts (float): The minimum count rate returned by this funtion.
+        """
+        try:
+            data = fits.getdata(filename)
+        except FileNotFoundError:
+            data = fits.getdata(filename + '.fz')
+
+        # Calculate average counts per pixel
+        mean_counts, _, _ = stats.sigma_clipped_stats(data-bias)
```

Comment:
@AnthonyHorton do we need the full frame for this or should it be from a cropped center? Not sure it matters much with the sigma-clipping though